tool/tsh/common/tsh_test.go

```diff
@@ -5422,3 +5422,61 @@ func Test_formatActiveDB(t *testing.T) {
 		})
 	}
 }
+
+func TestFlatten(t *testing.T) {
+	// Test setup: create a server and a user
+	home := t.TempDir()
+	identityPath := filepath.Join(t.TempDir(), "identity.pem")
+
+	alice, err := types.NewUser("alice@example.com")
+	require.NoError(t, err)
+	alice.SetRoles([]string{"access"})
+
+	connector := mockConnector(t)
+
+	authProcess, proxyProcess := makeTestServers(t, withBootstrap(connector, alice))
+	authServer := authProcess.GetAuthServer()
+	require.NotNil(t, authServer)
+
+	proxyAddr, err := proxyProcess.ProxyWebAddr()
+	require.NoError(t, err)
+
+	// Test setup: log in and obtain a valid identity for the user
+	conf := CLIConf{
+		Username:           alice.GetName(),
+		Proxy:              proxyAddr.String(),
+		InsecureSkipVerify: true,
+		IdentityFileOut:    identityPath,
+		IdentityFormat:     identityfile.FormatFile,
+		HomePath:           home,
+		AuthConnector:      connector.GetName(),
+		MockSSOLogin:       mockSSOLogin(t, authServer, alice),
+		Context:            context.Background(),
+	}
+	require.NoError(t, onLogin(&conf))
+
+	// Test setup: validate we got a valid identity
+	_, err = identityfile.KeyFromIdentityFile(identityPath, "proxy.example.com", "")
+	require.NoError(t, err)
+
+	// Test execution: flatten the identity previously obtained in a new home.
+	freshHome := t.TempDir()
+	conf = CLIConf{
+		Proxy:              proxyAddr.String(),
+		InsecureSkipVerify: true,
+		IdentityFileIn:     identityPath,
+		HomePath:           freshHome,
+		Context:            context.Background(),
+	}
+	require.NoError(t, onFlatten(&conf))
+
+	// Test execution: validate that the newly created profile can be used to build a valid client.
+	clt, err := makeClient(&conf)
+	require.NoError(t, err)
+
+	_, err = clt.Ping(context.Background())
+	require.NoError(t, err)
+
+	// Test execution: validate that flattening fails if a profile already exists.
+	require.Error(t, onFlatten(&conf), "expecting an error when overwriting an existing profile")
```

Comment:
This only fails because `onFlatten` sets `IdentityFileIn = ""` in the first call. In the current implementation, flatten actually succeeds in overwriting existing profiles. IMO this is preferable to erroring out and requiring the user to logout first. It also gives us the opportunity to rely on a current profile for missing cluster information.

This isn't quite working the way I expected though, which is my fault from the previous patch. Try this patch to fix it:

<details>
<summary>Patch</summary>
diff --git a/lib/client/identityfile/identity.go b/lib/client/identityfile/identity.go
index 1070abfc60..c9e669190b 100644
--- a/lib/client/identityfile/identity.go
+++ b/lib/client/identityfile/identity.go
@@ -835,7 +835,12 @@ func LoadIdentityFileIntoClientStore(store *client.Store, identityFile, proxyAdd
 		return trace.Wrap(err)
 	}
 
-	// Load the client key from the agent.
+	// This key may already exist in a tsh profile, delete it before overwriting
+	// it with the identity key to avoid leaving app/db/kube certs from the old key.
+	if err := store.DeleteKey(key.KeyIndex); err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+	}
+
 	if err := store.AddKey(key); err != nil {
 		return trace.Wrap(err)
 	}
diff --git a/tool/tsh/common/tsh.go b/tool/tsh/common/tsh.go
index ace05f118f..b6786dbc4b 100644
--- a/tool/tsh/common/tsh.go
+++ b/tool/tsh/common/tsh.go
@@ -4153,9 +4153,6 @@ func refuseArgs(command string, args []string) error {
 func onFlatten(cf *CLIConf) error {
 	// Save the identity file path for later
 	identityFile := cf.IdentityFileIn
-	if cf.Proxy == "" {
-		return trace.BadParameter("--proxy must be specified")
-	}
 
 	// We clear the identity flag so that the client store will be backed
 	// by the filesystem instead (in ~/.tsh or TELEPORT_HOME).
@@ -4167,6 +4164,11 @@ func onFlatten(cf *CLIConf) error {
 		return trace.Wrap(err)
 	}
 
+	// Proxy address may be loaded from existing tsh profile or from --proxy flag.
+	if c.WebProxyAddr == "" {
+		return trace.BadParameter("No proxy address specified, missed --proxy flag?")
+	}
+
 	// Load the identity file key and partial profile into the client store.
 	if err := identityfile.LoadIdentityFileIntoClientStore(c.ClientStore, identityFile, c.WebProxyAddr, c.SiteName); err != nil {
 		return trace.Wrap(err)
diff --git a/tool/tsh/common/tsh_test.go b/tool/tsh/common/tsh_test.go
index a8bbcda8c6..55bc046563 100644
--- a/tool/tsh/common/tsh_test.go
+++ b/tool/tsh/common/tsh_test.go
@@ -5477,6 +5477,7 @@ func TestFlatten(t *testing.T) {
 	_, err = clt.Ping(context.Background())
 	require.NoError(t, err)
 
-	// Test execution: validate that flattening fails if a profile already exists.
-	require.Error(t, onFlatten(&conf), "expecting an error when overwriting an existing profile")
+	// Test execution: validate that flattening succeeds if a profile already exists.
+	conf.IdentityFileIn = identityPath
+	require.NoError(t, onFlatten(&conf), "unexpected error when overwriting a tsh profile with tsh flatten")
 }
</details>

Reply:
applied